### DISCUSSION TITLE
Bring back "src" directory concept from Pymakr 1.x?

### DISCUSSION THREAD

**Discussion by VirtualWolf** [pycom/pymakr-vsc#264] [Ideas]
Pymakr 1.x had a setup where your actual Micropython source code lived in the `src` directory and when you uploaded to the device, only the files in that directory would be copied across, like this:

```
/
  - .micropy/
    - [...]
  - .vscode/
    - [...]
  - .pylintrc
  - README
  - src/ <-- Only the files under this directory are actually copied across to the device
      - boot.py
      - main.py
      - other_code.py
```

That way all the `.vscode`, `.micropy`, READMEs, and whatever other junk that lives in the root level of your repository wouldn't all be copied across as well. It'd be great to see Pymakr 2.x bring this back, because it's annoying having to wait for _all_ of the files to be copied across again when all I've changed is `main.py`. :)

**Comment by jakobrosenberg** [comment]
The preview version now has `dev mode` which automatically uploads changed files in realtime. In this mode, manually clicking the upload button will not reupload the entire project, but instead only upload files that are different from the existing.

Do you still think the source folder should be nested and if so, would something like this work for you?


```javascript
// pymakr.conf
{
  "projectDir":  "src"
}

**Reply by VirtualWolf** [reply]
That would be perfect! The dev mode and this aren't directly related in my mind, auto-uploading on save would be great but I still don't want to upload all my other random crap files that are in the root level of the project repository. 😁

**Reply by jakobrosenberg** [reply]
My original thought was to use something like `.pyignore` for that, but I like the idea of a `src` folder. One question though. How should the `pymakr.conf` be handled? I consider it the project manifest and would like it to travel with the project.

**Reply by VirtualWolf** [reply]
Not quite following what you're asking there I'm afraid...

**Reply by jakobrosenberg** [reply]
Currently in v2 the manifest `pymakr.conf` travels with the project. So if you sync a project to a device on machine A and then connect the device to machine B, you can then download the project with its manifest.

This probably shouldn't be a goal as I agree source files and dev environment should be kept separate from production code. That said, the `pymakr.conf` could in the future contain runtime info, but I don't know about that yet.

**Reply by VirtualWolf** [reply]
Oooh, I get you. To be honest, the thought of copying a project directory back from a device never occurred to me, I just assumed it'd be stored under version control on the computer so you'd just clone the repository on machine B!

**Comment by jakobrosenberg** [comment]
I'm working on this now, but have some questions about the name. The main issue is that we don't know if users are compiling their code before uploading it to the device. It would therefore be prudent to plan for nomenclature that works with both handwritten and compiled code.

I've listed some names below and attaches my thoughts.

|**Name**|**Issue**|
|-|-|
| **project_dir** | This is reserved for the root of the project, ie. the dirname of pymakr.json |
| **src** | Would be a misnomer for those who work with compiled code and it would conflict with the name of the user's actual source folder. |
| **dist_dir** | Distributable to one or more devices. Established convention. Can contain both source or compiled code, but has a connotation of compiled code. |
| **build_dir** | Doesn't work with source files. Same goes for `output_dir` etc. |
| **sync_dir** | Strong connotation of isomorphic bidirectional relationship between source files and device|
| **device_dir** | Slightly confusing as an option property, but `device` itself could make sense as an option value, eg. `dist_dir: './device'` |

What are your thoughts on this? Personally I'm inclined towards making this the new default for `pymakr.conf`:

```json
{
  "name": "My Project",
  "dist_dir": "./device"
}

**Reply by VirtualWolf** [reply]
`dist_dir` seems pretty reasonable to me!
